### ISSUE TITLE
Dynamic imports from CDN

### ISSUE THREAD

**Issue by hollowcat** [fuse-box/fuse-box#1556]
Hello. Experiencing problems attempting dynamic imports from CDN. Doing something wrong?

Using in index.js,

```
    await Promise.all([
        import("https://..."),
        import("https://..."), ...
    ]);
```

With cdnjs.cloudflare URLs, it gives

> Access to XMLHttpRequest at 'https://cdnjs.cloudflare.com/ajax/libs/...min.js' from origin 'http://127.0.0.1:8000' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource.

> OPTIONS 405 in the Network tab

Cloudflare has an example on how to simulate the OPTIONS method response via a service worker, but doing so didn't help.

With jsDelivr URLs, it gives

> Access to fetch at 'https://cdn.jsdelivr.net/npm/...min.js' from origin 'http://127.0.0.1:8000' has been blocked by CORS policy: Request header field x-requested-with is not allowed by Access-Control-Allow-Headers in preflight response.

> OPTIONS 200 in the Network tab

The errors occur for every import that I use. Everything works if imported via a script tag inside index.html.

However, it all _does work_ when using unpkg URLs.

Except a specific library I load, that has this init code

```
...else if (typeof module === 'object' && module.exports) {
    // Node/CommonJS
    module.exports = factory(require('jquery'), global, global.document);
  } else ...
```

and it alone gives

> Uncaught ReferenceError: require is not defined
>     at eval (eval at evaluateModule (app.js:), <anonymous>:)
>     at eval (eval at evaluateModule (app.js:), <anonymous>:)
>     at evaluateModule (app.js:)
>     at app.js:
>     at XMLHttpRequest.request.onreadystatechange (app.js:)

No, I tried adding shimming for jquery and it does not help.

Even with this one import commented out, it still does not work exactly like it does with script tag imports. Global objects that some libs define are missing when they are imported dynamically.

For clarity, what I want is a lean index.html with no script tags except the one for index.js. Index.js is to load my application after awaiting on imports of all of its dependencies from CDN.

A separate question: how do I unload a module when no longer needed to free up memory?

Thanks. Hoping too for SRI to be implemented in the next version.

**Comment by nchanged** [comment]
@hollowcat the problem with dynamic import from CDN is that we don't know which object is exposed and how it is exposed. Anyway, I am working on a brand new `v4` version which should address code splitting properly.

**Comment by nchanged** [comment]
Do you think we can just insert it to the head ? And resolve once loaded ?

**Comment by hollowcat** [comment]
I'm not qualified to provide technical suggestions, but my use case is with Lit-element (former Polymer, a.k.a. web components) + Redux + Redux-saga, as well as legacy jQuery code. My app's lean loader is to lazy-load code for components, sagas and store-related definitions, with the root component then assembling everything via dependency injection. Dynamic loading of CDN dependencies isn't critical, but nice for consistency.

Insert a CDN-sourced script tag to the head, with an onload event that resolves the import promise, like in [here](https://developer.mozilla.org/en-US/docs/Web/API/HTMLScriptElement#Dynamically_importing_scripts) ? Why not, and perhaps would bypass the aforementioned XHR CORS problem. No idea about exposed objects because I am not clear how module systems work.

**Comment by charsleysa** [comment]
@hollowcat can you please confirm if the issue still occurs with `fuse-box@4.0.0-next.400`?
Also see [here for info on major changes](https://github.com/fuse-box/fuse-box/blob/master/docs/changelog-fuse-compiler.md)

**Comment by hollowcat** [comment]
@charsleysa 

I've just updated to next.401 and modified my fuse.js as written (replaced `output` with `bundles` as an argument to `runDev`, no other changes from [this example](https://github.com/fuse-box/react-example/blob/v4/fuse.ts)).

It does compile, but gives _TypeError: Object.assign(...) is not a function_ when opened in the browser, generated at the point of first import of LitElement. My config has the `dependencies: { include: [ 'tslib' ] }` line, my tsconfig.js has `importHelpers: true`, and it previously worked.

So I can't report on the CDN imports issue until I figure out how to just make it work again.

**Comment by nchanged** [comment]
Could you reproduce it ?

**Comment by nchanged** [comment]
You don't need to add tslib anymore btw, we don't use Typescript anymore

**Comment by hollowcat** [comment]
With only this code in my entrypoint index.ts,

```
(async () => {
    await import("https://unpkg.com/redux@4.0.1/dist/redux.min.js"); // works in 164
    await import("https://cdn.jsdelivr.net/npm/redux@4.0.1/dist/redux.min.js");
    await import("https://cdnjs.cloudflare.com/ajax/libs/redux/4.0.1/redux.min.js");
})();
```
next.164 throws as described initially, while next.403 just throws _Uncaught (in promise) Error: Module ...redux@4.0.1/dist/redux.min.js was not found_ any way.


Does the shift to another parser affect compatibility with Lit-element which expects TS?

How to fix the _Object.assign is not a function_ errors?

**Comment by nchanged** [comment]
Are you using IE7 or something? Can you add a polyfill ?

Also did you previously use any ts transformers ? Cuz the latest version requires custom transformers, we are trying to cover the most popular cases but it will take time.

As for dynamic cdn imports I will add the support for that soon enough

**Comment by hollowcat** [comment]
Using latest FF and Chrome, no transformers or Babel, adding the polyfill [from here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign#Polyfill) did not help.

The _Object.assign_ issue first appears with fusebox version alpha.269; version alpha.267 and some before it throw a different error, _File.js does not provide a module_; versions between that and next.164 generate bundles with various syntax errors.

**Comment by charsleysa** [comment]
@hollowcat I've been able to successfully run the [fuse-box react example project v4 branch](https://github.com/fuse-box/react-example/blob/v4/fuse.ts).

The only changes required were:
- delete the package-lock.json file
- update `fuse.ts` as the `watch` option has been renamed to `watcher`
- update `package.json` and change `fuse-box` to `4.0.0-next.403`

If you are still having issues running that project please open a new issue as this issue will be specifically for dynamic url imports.

**Comment by charsleysa** [comment]
Closed due to inactivity.
